File: db/assignment.go
// Package db 負責場所（venues）與指派（assignments），對齊討論 001 身份與職業分離。
package db

import (
	"singularity_world/store"
)

// Venue 場所：id、名稱、room_ids（JSON 陣列）。
type Venue struct {
	ID      string
	Name    string
	RoomIDs []string
}

// Assignment 指派：誰、什麼職業、哪個場所。
type Assignment struct {
	EntityID     string
	OccupationID string
	VenueID      string
	AssignedBy   string
}

// SeedVenues 確保預設場所存在；現由 data/venues.json 於 store.Init 載入，此函式為 no-op。
func SeedVenues() error {
	return nil
}

// InsertAssignment 新增一筆指派；已存在則忽略。
func InsertAssignment(entityID, occupationID, venueID, assignedBy string) error {
	if store.Default == nil {
		return ErrNoStore
	}
	return store.Default.InsertAssignment(entityID, occupationID, venueID, assignedBy)
}

// GetAssignmentsForEntity 取得某實體的全部指派（職業＋場所）。
func GetAssignmentsForEntity(entityID string) ([]Assignment, error) {
	if store.Default == nil {
		return nil, ErrNoStore
	}
	slist := store.Default.GetAssignmentsForEntity(entityID)
	list := make([]Assignment, len(slist))
	for i, a := range slist {
		list[i] = Assignment{
			EntityID: a.EntityID, OccupationID: a.OccupationID, VenueID: a.VenueID,
			AssignedBy: a.AssignedBy,
		}
	}
	return list, nil
}

// GetNPCTitleFromAssignments 依指派推導職稱；無指派則回傳空字串。
func GetNPCTitleFromAssignments(entityID string) string {
	list, err := GetAssignmentsForEntity(entityID)
	if err != nil || len(list) == 0 {
		return ""
	}
	return list[0].OccupationID
}

// RemoveAssignmentsForEntity 移除某實體的全部指派（死亡除名用）。
func RemoveAssignmentsForEntity(entityID string) error {
	if store.Default == nil {
		return ErrNoStore
	}
	return store.Default.RemoveAssignmentsForEntity(entityID)
}

// GetVenueMaxStaff 回傳該場所職缺上限；0 用 defaultMax。
func GetVenueMaxStaff(venueID string, defaultMax int) int {
	if store.Default == nil {
		return defaultMax
	}
	return store.Default.GetVenueMaxStaff(venueID, defaultMax)
}

// GetFirstOccupationIDForVenue 回傳該場所既有指派中的第一個職業 ID。
func GetFirstOccupationIDForVenue(venueID string) string {
	if store.Default == nil {
		return ""
	}
	return store.Default.GetFirstOccupationIDForVenue(venueID)
}

// GetRoomIDsForVenue 回傳該場所涵蓋的房間 ID 列表。
func GetRoomIDsForVenue(venueID string) ([]string, error) {
	if store.Default == nil {
		return nil, ErrNoStore
	}
	v := store.Default.GetVenue(venueID)
	if v == nil {
		return nil, nil
	}
	return v.RoomIDs, nil
}

// IsRoomInVenue 判斷房間是否在該場所的 room_ids 內。
func IsRoomInVenue(roomID, venueID string) (bool, error) {
	if store.Default == nil {
		return false, ErrNoStore
	}
	return store.Default.IsRoomInVenue(roomID, venueID), nil
}

// EntityInVenueAtRoom 回傳該實體在指定房間時，是否處於任一指派場所內。
func EntityInVenueAtRoom(entityID, roomID string) (bool, error) {
	list, err := GetAssignmentsForEntity(entityID)
	if err != nil {
		return false, err
	}
	for _, a := range list {
		ok, err := IsRoomInVenue(roomID, a.VenueID)
		if err != nil {
			return false, err
		}
		if ok {
			return true, nil
		}
	}
	return false, nil
}

// GetVenueIDsForRoom 回傳包含該房間的所有場所 ID。
func GetVenueIDsForRoom(roomID string) ([]string, error) {
	if store.Default == nil {
		return nil, ErrNoStore
	}
	return store.Default.GetVenueIDsForRoom(roomID), nil
}

// GetAssignmentCountByVenue 回傳該場所目前的指派數量。
func GetAssignmentCountByVenue(venueID string) (int, error) {
	if store.Default == nil {
		return 0, ErrNoStore
	}
	return store.Default.GetAssignmentCountByVenue(venueID), nil
}

// GetAllVenueIDs 回傳所有場所 ID。
func GetAllVenueIDs() ([]string, error) {
	if store.Default == nil {
		return nil, ErrNoStore
	}
	return store.Default.GetAllVenueIDs(), nil
}

// GetAllVenueRoomIDs 回傳所有場所涵蓋的房間 ID。
func GetAllVenueRoomIDs() ([]string, error) {
	if store.Default == nil {
		return nil, ErrNoStore
	}
	return store.Default.GetAllVenueRoomIDs(), nil
}

File: src/db/assignment.rs
// 場所與指派門面，對齊既有 db/assignment。

pub use crate::store::{Assignment, Venue};

use crate::store;

use super::ErrNoStore;

/// 確保預設場所存在；現由 `venues.json` 於 store 初始化載入，此函式為 no-op（對齊既有 `SeedVenues`）。
pub fn seed_venues() -> anyhow::Result<()> {
    Ok(())
}

/// 新增一筆指派；同 entity+occupation+venue 已存在則忽略（對齊既有 `InsertAssignment`）。
pub fn insert_assignment(
    entity_id: &str,
    occupation_id: &str,
    venue_id: &str,
    assigned_by: &str,
) -> anyhow::Result<()> {
    let arc = store::get_store().ok_or(ErrNoStore)?;
    let mut s = arc.write().unwrap();
    s.insert_assignment(entity_id, occupation_id, venue_id, assigned_by)
}

/// 某實體的全部指派（對齊既有 `GetAssignmentsForEntity`）。
pub fn get_assignments_for_entity(entity_id: &str) -> anyhow::Result<Vec<Assignment>> {
    let arc = store::get_store().ok_or(ErrNoStore)?;
    let s = arc.read().unwrap();
    Ok(s.get_assignments_for_entity(entity_id))
}

/// 依指派推導職稱；無指派則空字串（對齊既有 `GetNPCTitleFromAssignments`）。
#[must_use]
pub fn get_npc_title_from_assignments(entity_id: &str) -> String {
    let Ok(list) = get_assignments_for_entity(entity_id) else {
        return String::new();
    };
    list.first().map(|a| a.occupation_id.clone()).unwrap_or_default()
}

/// 移除某實體的全部指派（對齊既有 `RemoveAssignmentsForEntity`）。
pub fn remove_assignments_for_entity(entity_id: &str) -> anyhow::Result<()> {
    let arc = store::get_store().ok_or(ErrNoStore)?;
    let mut s = arc.write().unwrap();
    s.remove_assignments_for_entity(entity_id)
}

/// 該場所職缺上限；0 用 `default_max`（對齊既有 `GetVenueMaxStaff`）。
#[must_use]
pub fn get_venue_max_staff(venue_id: &str, default_max: i32) -> i32 {
    let Some(arc) = store::get_store() else {
        return default_max;
    };
    let s = arc.read().unwrap();
    s.get_venue_max_staff(venue_id, default_max)
}

/// 該場所既有指派中的第一個職業 ID（對齊既有 `GetFirstOccupationIDForVenue`）。
#[must_use]
pub fn get_first_occupation_id_for_venue(venue_id: &str) -> String {
    let Some(arc) = store::get_store() else {
        return String::new();
    };
    let s = arc.read().unwrap();
    s.get_first_occupation_id_for_venue(venue_id)
}

/// 該場所涵蓋的房間 ID（對齊既有 `GetRoomIDsForVenue`）。
pub fn get_room_ids_for_venue(venue_id: &str) -> anyhow::Result<Option<Vec<String>>> {
    let arc = store::get_store().ok_or(ErrNoStore)?;
    let s = arc.read().unwrap();
    Ok(s.get_venue(venue_id).map(|v| v.room_ids.clone()))
}

/// 房間是否在該場所內（對齊既有 `IsRoomInVenue`）。
pub fn is_room_in_venue(room_id: &str, venue_id: &str) -> anyhow::Result<bool> {
    let arc = store::get_store().ok_or(ErrNoStore)?;
    let s = arc.read().unwrap();
    Ok(s.is_room_in_venue(room_id, venue_id))
}

/// 實體在指定房間時是否處於任一指派場所內（對齊既有 `EntityInVenueAtRoom`）。
pub fn entity_in_venue_at_room(entity_id: &str, room_id: &str) -> anyhow::Result<bool> {
    for a in get_assignments_for_entity(entity_id)? {
        if is_room_in_venue(room_id, &a.venue_id)? {
            return Ok(true);
        }
    }
    Ok(false)
}

/// 包含該房間的所有場所 ID（對齊既有 `GetVenueIDsForRoom`）。
pub fn get_venue_ids_for_room(room_id: &str) -> anyhow::Result<Vec<String>> {
    let arc = store::get_store().ok_or(ErrNoStore)?;
    let s = arc.read().unwrap();
    Ok(s.get_venue_ids_for_room(room_id))
}

/// 該場所目前的指派數量（對齊既有 `GetAssignmentCountByVenue`）。
pub fn get_assignment_count_by_venue(venue_id: &str) -> anyhow::Result<usize> {
    let arc = store::get_store().ok_or(ErrNoStore)?;
    let s = arc.read().unwrap();
    Ok(s.get_assignment_count_by_venue(venue_id))
}

/// 所有場所 ID（對齊既有 `GetAllVenueIDs`）。
pub fn get_all_venue_ids() -> anyhow::Result<Vec<String>> {
    let arc = store::get_store().ok_or(ErrNoStore)?;
    let s = arc.read().unwrap();
    Ok(s.get_all_venue_ids())
}

/// 所有場所涵蓋的房間 ID（對齊既有 `GetAllVenueRoomIDs`）。
pub fn get_all_venue_room_ids() -> anyhow::Result<Vec<String>> {
    let arc = store::get_store().ok_or(ErrNoStore)?;
    let s = arc.read().unwrap();
    Ok(s.get_all_venue_room_ids())
}
